server.go

```diff
@@ -45,10 +45,20 @@ func NewServer() *Server {
 	}
 }
 
+// RequestInfo contains all the information we pass to before/after functions
+type RequestInfo struct {
+	Method     string
+	Error      error
+	Request    *http.Request
+	StatusCode int
+}
+
 // Server serves registered RPC services using registered codecs.
 type Server struct {
-	codecs   map[string]Codec
-	services *serviceMap
+	codecs     map[string]Codec
+	services   *serviceMap
+	beforeFunc *func(i *RequestInfo)
```

Comment:
I don't think these need to be pointer types. Just `func (i *RequestInfo)` should work since the value of a func is just a pointer anyway.